core/arch/arm/sm/suspend.S

```diff
@@ -0,0 +1,224 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/unwind.h>
+#include <asm.S>
+#include <asm-defines.h>
+#include <arm.h>
+#include <arm32_macros.S>
+#include <arm32_macros_cortex_a9.S>
+#include <kernel/asan.h>
+#include <platform_config.h>
+#include <keep.h>
+#include <kernel/unwind.h>
+
+.section .text
+FUNC cpu_suspend, :
+UNWIND( .fnstart)
+	push	{r4 - r12, lr}
+UNWIND(.save	{r4 - r12, lr})
+	mov	r5, sp
+	sub	sp, sp, #SM_PM_CTX_SIZE
+	push	{r0, r1}
+
+	mov	r2, r5
+	mov	r1, #SM_PM_CTX_SIZE
+	add	r0, sp, #8
+	blx	cpu_suspend_save
+	adr	lr, aborted
+	/* Jump to arch specific suspend*/
+	pop	{r0, pc}
+aborted:
+	/* cpu not suspended */
+	add	sp, sp, #SM_PM_CTX_SIZE
+	/* Return -1 to the caller */
+	mov	r0, #(-1)
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC cpu_suspend
+
+FUNC cpu_v7_suspend, :
+UNWIND(.fnstart)
+	push	{r4 - r12, lr}
+	read_fcseidr r4
+	read_tpidruro r5
+	stmia	r0!, {r4 - r5}
+	read_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported"
+#else
+	read_ttbr0 r5
+	read_ttbr1 r6
+	read_ttbcr r7
+#endif
+	read_sctlr r8
+	read_actlr r9
+	read_cpacr r10
+	read_mvbar r11
+	stmia	r0!, {r4 - r11}
+	read_prrr r4
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC cpu_v7_suspend
+
+FUNC cpu_do_suspend, :
+UNWIND(.fnstart)
+	push	{r4, r5}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr     r4, =A7_PART_NUM
+	cmp	r5, r4
+	beq	a7_suspend
+	/*
+	 * TODO: add more Cortex A[x] check?
+	 *
+	 * A9 needs PCR/DIAG
+	 */
+	read_pcr r4
+	read_diag r5
+	stmia	r0!, {r4 - r5}
+a7_suspend:
+	pop	{r4, r5}
+	b	cpu_v7_suspend
+UNWIND( .fnend)
+END_FUNC cpu_do_suspend
+
+FUNC cpu_resume, :
+UNWIND(.fnstart)
+	cpsid	aif
+
+	adr	r0, _suspend_sp
+	ldr	r2, [r0]
+	/* Get the physical address of thread_core_local */
```

Comment:
Let's skip preparing for non-linear mapping here to focus on one issue at a time.